File: macos/Sources/VoiceTyper/ASR/RecognitionBuffer.swift
import Foundation

/// 单次录音会话的音频累积与滑窗预览。移植自服务端 `recognizer.SenseVoiceSession`
/// （见 macos/DESIGN.md §5.2），逻辑原样搬运，不做"改进"：
///
/// SenseVoice 本身非流式，预览靠"对已累积音频整段重跑"实现。但对全部音频重跑的成本
/// 随录音时长线性增长而预览频率固定，于是单会话累计 CPU 会随录音时长呈平方增长。
/// 用一条滚动窗口规避：窗口左侧的音频只识别一次、固化进 `committedText`，之后的预览
/// 只对窗口右侧（最近 `previewWindowSamples` 个采样点）重跑。
///
/// 每次新建、用完即弃；非线程安全的部分（`committedText`/`committedN`）只应在
/// `ASRService.asrQueue` 上访问；`append` 允许从任意线程调用（内部用锁保护）。
final class RecognitionBuffer: @unchecked Sendable {
    /// 预览窗口（采样点，16kHz）。稳态预览 CPU ≈ RTF × 15s / 0.6s（0.6s 为客户端送帧间隔）。
    static let previewWindowSamples = 15 * 16000
    /// 窗口滚动时，在目标切点 ±100ms 内挑能量最低的 10ms 处下刀。
    private static let seamSearchRadius = 1600
    private static let seamStep = 160

    private let engine: any SenseVoiceRecognizing
    private let lock = NSLock()
    private var chunks: [[Float]] = []
    private var n = 0
    private var joined: [Float]?

    /// 窗口左侧已固化的预览文本；只在 asrQueue 上读写。
    private var committedText = ""
    private var committedN = 0

    init(engine: any SenseVoiceRecognizing) {
        self.engine = engine
    }

    /// 累计接收的样本数，供调用方做会话时长上限判断。可从任意线程调用。
    var sampleCount: Int {
        lock.lock()
        defer { lock.unlock() }
        return n
    }

    /// 累积音频。必须足够便宜——可能跑在音频采集的回调路径上。
    func append(_ chunk: [Float]) {
        guard !chunk.isEmpty else { return }
        lock.lock()
        chunks.append(chunk)
        n += chunk.count
        joined = nil
        lock.unlock()
    }

    /// 返回全量预览文本 = 已固化前缀 + 当前窗口的识别结果。只应在 asrQueue 上调用。
    func preview() throws -> String {
        let audio = audioBuffer()
        if audio.count - committedN > Self.previewWindowSamples {
            try roll(audio)
        }
        let tail = Array(audio[committedN...])
        return committedText + (try engine.recognize(tail))
    }

    /// 松手后调用：对完整音频重新整段识别。刻意不复用 preview 的拼接结果——
    /// 窗口边界处的固化文本可能有接缝瑕疵，finalize 产出的是要上屏的最终文本，
    /// 必须整段重跑。只应在 asrQueue 上调用。
    func finalize() throws -> String {
        try engine.recognize(audioBuffer())
    }

    /// 把窗口前半段识别成文本固化，之后的预览不再重复识别这段音频。
    private func roll(_ audio: [Float]) throws {
        let target = committedN + Self.previewWindowSamples / 2
        let cut = Self.findSeam(audio, target: target)
        let segment = Array(audio[committedN..<cut])
        committedText += try engine.recognize(segment)
        committedN = cut
    }

    /// 在 target 附近找能量（RMS）最低的一小段作为切点，尽量避免把一个字切在正中间。
    private static func findSeam(_ audio: [Float], target: Int) -> Int {
        let lo = max(0, target - seamSearchRadius)
        let hi = min(audio.count, target + seamSearchRadius)
        let span = ((hi - lo) / seamStep) * seamStep
        guard span > 0 else { return target }

        var bestEnergy = Float.greatestFiniteMagnitude
        var bestOffset = 0
        var offset = 0
        while offset < span {
            var sum: Float = 0
            for i in 0..<seamStep {
                let sample = audio[lo + offset + i]
                sum += sample * sample
            }
            let energy = sum / Float(seamStep)
            if energy < bestEnergy {
                bestEnergy = energy
                bestOffset = offset
            }
            offset += seamStep
        }
        return lo + bestOffset
    }

    private func audioBuffer() -> [Float] {
        lock.lock()
        if let joined {
            lock.unlock()
            return joined
        }
        guard !chunks.isEmpty else {
            lock.unlock()
            return []
        }
        var flat = [Float]()
        flat.reserveCapacity(n)
        for chunk in chunks { flat.append(contentsOf: chunk) }
        joined = flat
        lock.unlock()
        return flat
    }
}

File: windows/Asr/RecognitionBuffer.cs
using System;

namespace VoiceTyper.Asr;

/// <summary>
/// 单次录音会话的音频累积与滑窗预览。C# 直译自
/// <c>macos/Sources/VoiceTyper/ASR/RecognitionBuffer.swift</c>（移植自服务端
/// <c>recognizer.SenseVoiceSession</c>），逻辑原样搬运，不做"改进"：
///
/// SenseVoice 本身非流式，预览靠"对已累积音频整段重跑"实现。用一条滚动窗口规避成本随
/// 录音时长平方增长：窗口左侧的音频只识别一次、固化进 <see cref="_committedText"/>，
/// 之后的预览只对窗口右侧（最近 <c>previewWindowSamples</c> 个采样点）重跑。
///
/// 每次新建、用完即弃；<c>_committedText</c>/<c>_committedN</c> 只应在 <see cref="AsrPump"/> 线程上
/// 访问；<see cref="Append"/> 允许从任意线程调用（内部用 <c>lock</c> 保护）。
///
/// 单条持续增长的扁平缓冲：旧实现是 <c>List&lt;float[]&gt; _chunks</c> + <c>_joined</c> 缓存，
/// <c>Append</c> 每次都会让 <c>_joined</c> 失效——预览节奏固定在约 600ms 一次，但每次都要把
/// <b>全部</b>已累积音频重新拼接一遍，开销与总录音时长成正比（R3-08）。改成单条只增不改的
/// 扁平数组后，<see cref="Append"/> 退化成均摊 O(chunk 大小) 的操作；读取只需要在锁内快照
/// <c>(_buffer 引用, _count)</c> 这一对值——扩容只会替换 <c>_buffer</c> 字段指向一个新数组，
/// 从不就地修改旧数组的已提交内容，因此持有旧引用的读者始终能安全读到自己快照时刻的数据，
/// 不需要再额外复制一份（macOS 侧靠 Swift Array 的写时复制免费获得同样的效果）。
/// </summary>
internal sealed class RecognitionBuffer
{
    /// <summary>窗口滚动时，在目标切点 ±100ms 内挑能量最低的 10ms 处下刀。</summary>
    private const int SeamSearchRadius = 1600;
    private const int SeamStep = 160;

    private readonly ISenseVoiceRecognizing _engine;
    private readonly int _previewWindowSamples;
    private readonly object _lock = new();
    private float[] _buffer = Array.Empty<float>();
    private int _count;

    /// <summary>窗口左侧已固化的预览文本；只在 AsrPump 线程上读写。</summary>
    private string _committedText = "";
    private int _committedN;

    /// <param name="previewWindowSamples">
    /// 预览窗口（采样点，16kHz）。默认 15 秒，与服务端/macOS 一致；
    /// Windows 侧允许按本机实测 RTF 调小（见 windows/DESIGN.md §4.5 的自校准）。
    /// </param>
    public RecognitionBuffer(ISenseVoiceRecognizing engine, int previewWindowSamples = 15 * 16000)
    {
        _engine = engine;
        _previewWindowSamples = previewWindowSamples;
    }

    /// <summary>累计接收的样本数，供调用方做会话时长上限判断。可从任意线程调用。</summary>
    public int SampleCount
    {
        get { lock (_lock) return _count; }
    }

    /// <summary>累积音频。必须足够便宜——可能跑在音频采集的回调路径上。均摊 O(chunk 大小)。</summary>
    public void Append(float[] chunk)
    {
        if (chunk.Length == 0) return;
        lock (_lock)
        {
            EnsureCapacityLocked(_count + chunk.Length);
            Array.Copy(chunk, 0, _buffer, _count, chunk.Length);
            _count += chunk.Length;
        }
    }

    /// <summary>返回全量预览文本 = 已固化前缀 + 当前窗口的识别结果。只应在 AsrPump 线程上调用。</summary>
    public string Preview()
    {
        var (buffer, count) = Snapshot();
        if (count - _committedN > _previewWindowSamples)
        {
            Roll(buffer, count);
        }
        var tail = buffer.AsSpan(_committedN, count - _committedN);
        return _committedText + _engine.Recognize(tail);
    }

    /// <summary>
    /// 松手后调用：对完整音频重新整段识别。刻意不复用 Preview 的拼接结果——窗口边界处的固化
    /// 文本可能有接缝瑕疵，Finalize 产出的是要上屏的最终文本，必须整段重跑。只应在 AsrPump 线程上调用。
    /// </summary>
    public string Finalize()
    {
        var (buffer, count) = Snapshot();
        return _engine.Recognize(buffer.AsSpan(0, count));
    }

    /// <summary>把窗口前半段识别成文本固化，之后的预览不再重复识别这段音频。</summary>
    private void Roll(float[] buffer, int count)
    {
        int target = _committedN + _previewWindowSamples / 2;
        int cut = FindSeam(buffer, count, target);
        var segment = buffer.AsSpan(_committedN, cut - _committedN);
        _committedText += _engine.Recognize(segment);
        _committedN = cut;
    }

    /// <summary>在 target 附近找能量（RMS）最低的一小段作为切点，尽量避免把一个字切在正中间。</summary>
    private static int FindSeam(float[] buffer, int count, int target)
    {
        int lo = Math.Max(0, target - SeamSearchRadius);
        int hi = Math.Min(count, target + SeamSearchRadius);
        int span = (hi - lo) / SeamStep * SeamStep;
        if (span <= 0) return target;

        float bestEnergy = float.MaxValue;
        int bestOffset = 0;
        for (int offset = 0; offset < span; offset += SeamStep)
        {
            float sum = 0;
            for (int i = 0; i < SeamStep; i++)
            {
                float sample = buffer[lo + offset + i];
                sum += sample * sample;
            }
            float energy = sum / SeamStep;
            if (energy < bestEnergy)
            {
                bestEnergy = energy;
                bestOffset = offset;
            }
        }
        return lo + bestOffset;
    }

    /// <summary>必须在持有 <see cref="_lock"/> 时调用。按需把 <see cref="_buffer"/> 换成一个更大的
    /// 新数组（旧数组内容原样拷贝到新数组前缀，旧数组本身不再被 <see cref="_buffer"/> 引用，
    /// 但仍可能被并发读者的快照持有——绝不就地修改旧数组，见类文档）。</summary>
    private void EnsureCapacityLocked(int minCapacity)
    {
        if (_buffer.Length >= minCapacity) return;
        int newCapacity = Math.Max(minCapacity, Math.Max(_buffer.Length * 2, 16384));
        var grown = new float[newCapacity];
        Array.Copy(_buffer, grown, _count);
        _buffer = grown;
    }

    private (float[] Buffer, int Count) Snapshot()
    {
        lock (_lock) return (_buffer, _count);
    }
}
